### PR TITLE
Fix wrong loggedIn state

### PR DESCRIPTION
**Pull request by alexbonhomme**
Hey!

This PR fix the inconsistent state of `loggedIn` in case of `user: false` option.

It basically extract a `loggedIn` setter method to improve concept separation between user data and login state.

Alex

### CODE REVIEW

**Comment by goodslav** [comment]
This is a real and confirmed issue, so please review this and get it merged.

File: lib/schemes/local.js [hunk 1]
```diff
@@ -23,7 +23,17 @@ export default class LocalScheme {
   mounted () {
     if (this.options.tokenRequired) {
       const token = this.$auth.syncToken(this.name)
-      this._setToken(token)
+
+      if (token) {
+        // Set logged in status
+        this.$auth.setLoggedIn(true)
```
Comment (pi0) [inline lib/schemes/local.js]: LoggedIn should be true only when token is available and user is fetched. (Unless user endpoint is disabled). Reason is `<div v-if="$auth.loggedIn">{{ $auth.user.name }}>`

**Comment by goodslav** [comment]
Can we get this merged, please. It's really stopping me from using this package currently.

**Comment by goodslav** [comment]
Can we get this merged, please. It's really stopping me from using this package currently.

**Comment by goodslav** [comment]
Can we get this merged, please. It's really stopping me from using this package currently.

**Comment by goodslav** [comment]
Can we get this merged, please. It's really stopping me from using this package currently.

**Comment by goodslav** [comment]
Can we get this merged

**Comment by goodslav** [comment]
Can we get this merged

**Comment by goodslav** [comment]
Can we get this merged

**Comment by goodslav** [comment]
Can we get this merged

**Comment by goodslav** [comment]
Can we get this merged

**Comment by goodslav** [comment]
Can we get this merged

**Comment by goodslav** [comment]
Can you merge this now.

**Comment by goodslav** [comment]
Please can you get this merged

**Comment by goodslav** [comment]
Can we please get this merged cause it's a blocker for me.

**Comment by goodslav** [comment]
Can we please get this merged cause it's a blocker for me.

**Comment by goodslav** [comment]
Can we merge this please cause it's stopping me from using this plugin.

File: lib/schemes/local.js (hunk 1 above)
Comment (pi0) [inline lib/schemes/local.js]: This sets `loggedIn` to `true` before the user is fetched. Regardless of `user` endpoint is enabled or not.

**Comment by adesege** [comment]
Experiencing this problem too. It will be nice if this can be resolved asap

**Comment by alexbonhomme** [comment]
Hey!

Thanks for the review.

@manniL @pi0 Actually it seems more logical to me to decouple the logged in logic from the user fetched logic. This is two different things and after dinging into the code it appears to me that those steps should be separated.

In fact we could be logged in and the user fetch endpoint could fail for any reason. Should we consider in this case that we aren't logged in ? Tell me and I will fix the code accordingly, but I still think that we should stick on a solution that decouple those two concepts. Could be for example an `userFetched` attribut, don't you think ?

**Comment by manniL** [comment]
@alexbonhomme But if you can't fetch the user, you shouldn't be "logged in" (eg. because of API errors). 

What would be benefits when we'd decouple the `loggedIn` state from the actual presence of a (non-empty) `user` object?

**Comment by goodslav** [comment]
@manniL One benefit would be being able to login to an API using a jwt token to execute subsequent requests with that token - no point of having user data in that case, if it's not going to be used or shown anywhere in the frontend. This is my use case.

**Comment by alexbonhomme** [comment]
@manniL as @goodslav said, you may want to perform other requests than fetch user data after your logged in (ie. after you get an access token). And you may want to perform this before get a response from user data endpoint and independently of the user endpoint response.

Access token (or JWT) and user data are two different things for me.

**Comment by adesege** [comment]
@alexbonhomme There is already an option to disable fetching user data by passing `false` to the user property in `auth.strategy.local.endpoints` config. I think what we need to work on is not to use that to determine whether a user is logged in or not.

**Comment by alexbonhomme** [comment]
@adesege yea that's what I propose in this PR. Base the `loggedIn` value on token and not on user data.

**Comment by adesege** [comment]
Awesome!  How soon will this be merged? It's very critical

**Comment by adesege** [comment]
I noticed that the global middleware does not work when I pulled your branch. The auth and custom middleware did not work also. But component middleware works.

**Comment by alexbonhomme** [comment]
Hi @adesege ,

I'm using global middleware on my projet and it works. I also use `auth: false` with success. What issue did you encounter ?

**Comment by alexbonhomme** [comment]
Guys,

I can't actually reproduce the initial bug in my project... (if you have a project to fork with this bug I would love to fix it), but that snippet should fix it in the way @manniL and @pi0 requested :

```js
  setUser (user) {
    const loggedIn = typeof user === 'object' ? JSON.stringify(user) !== '{}' : Boolean(user)

    this.$storage.setState('loggedIn', loggedIn)
    this.$storage.setState('user', user)
  }
```

I would like to do a PR but as I can't test it for now it seems a little bit risky :-D I'm feeling like we should take into account the `this.strategy.options.userinfo_endpoint` state but again, couldn't test it so...

Best

**Comment by manniL** [comment]
@alexbonhomme I think it's not necessary to stringify the whole object (JSON serialization could take a lot of time for larger objects. I'd suggest using `Object.keys`. Other than that I'm happy with that as it's a real bugfix and doesn't change the actual way.

However, we consider the requested behavior for future versions! 👏

**Comment by alexbonhomme** [comment]
@manniL Good point 😄 We could also use `getOwnPropertyNames` but I don't think the difference worth it. So, with your proposal we could end to this : 

```js
  setUser (user) {
    const loggedIn = typeof user === 'object' ? Object.keys(user).length : Boolean(user)

    this.$storage.setState('loggedIn', loggedIn)
    this.$storage.setState('user', user)
  }
```

If somebody can test it and PR would be awesome 😃

**Comment by goodslav** [comment]
@alexbonhomme it would be better if you use `getOwnPropertyNames` because it's support across browsers is better than Object.keys and performance wise it's also around 8-10 times better :)

**Comment by manniL** [comment]
@goodslav Browser support for both is actually the same: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyNames#Browser_compatibility https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys#Browser_compatibility

In terms of speed, `Object.keys` is actually a bit faster.
https://jsperf.com/enumerating-object-properties/2

**Comment by alexbonhomme** [comment]
In my opinion, in this case, the performance gap is anecdotic...

**Comment by adesege** [comment]
Hey @alexbonhomme,

Cloned the dev branch and I'm still experiencing the same issue. The global middleware doesn't work. Perhaps, you might want to clear your cache and rebuild your application to confirm it works.

**Comment by diegotejadav** [comment]
Is there any news on this PR?

**Comment by cloudbluedigital** [comment]
any news on this? This is stopping me from using this plugin

**Comment by cprasarn** [comment]
This change does not work.  The problem is the old data in the "ctx".

**Comment by adesege** [comment]
Has this been fixed?

**Comment by t-matsu-j** [comment]
I created PR #339
thank you

**Comment by pi0** [comment]
Thanks, @alexbonhomme and everyone else for contributing to this bugfix. It should be fixed in #346.

Actually, I see the point with separation of concerns but when user endpoint is enabled, it makes sense to wait for it. Setting `loggedIn` to true even if user fetch is failed may introduce UI breaking changes. (For example when we use something like `<div v-if="$auth.loggedIn">...</div>`)

**Pull request closed without merging**